### PR TITLE
feat(core): add support for httpProxy and httpsProxy in networkSettings

### PR DESCRIPTION
**Pull request by merceyz**
**What's the problem this PR addresses?**

It's only possible to set a proxy for all traffic or no traffic.

Fixes https://github.com/yarnpkg/berry/issues/1531
Fixes https://github.com/yarnpkg/berry/issues/1530
Closes https://github.com/yarnpkg/berry/pull/2104

**How did you fix it?**

Added support for setting `httpProxy` and `httpsProxy` in network settings

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-core/sources/Configuration.ts [hunk 1]
```diff
@@ -345,6 +345,16 @@ export const coreDefinitions: {[coreSettingName: string]: SettingsDefinition} =
...
         },
+        httpProxy: {
+          description: `URL of the http proxy that must be used for outgoing http requests`,
+          type: SettingsType.STRING,
+          default: process.env.HTTP_PROXY || process.env.http_proxy || null,
+        },
+        httpsProxy: {
+          description: `URL of the http proxy that must be used for outgoing https requests`,
+          type: SettingsType.STRING,
+          default: process.env.HTTPS_PROXY || process.env.https_proxy || null,
```
Comment (bgotink) [inline packages/yarnpkg-core/sources/Configuration.ts:356]: Shouldn't these two default to `null`?
Right now, a configuration like

```yaml
httpsProxy: 'https://proxy.local'

networkSettings:
  custom-ca.local:
    caFilePath: /etc/custom-ca.cert
```

will make requests to `custom-ca.local` use the proxy configured via `HTTPS_PROXY`—if present in the environment—while all other domains will use the configured `httpsProxy`. If `HTTPS_PROXY` is not set in the environment, the configured `httpsProxy` will be used for all requests.

Comment (merceyz) [reply packages/yarnpkg-core/sources/Configuration.ts]: Indeed they should, cheers

File: packages/yarnpkg-core/sources/httpUtils.ts [hunk 2]
```diff
@@ -47,32 +50,50 @@ export function getNetworkSettings(target: string, opts: { configuration: Config
...
-    mergedNetworkSettings.caFilePath = opts.configuration.get(`caFilePath`);
+  // Apply defaults
+  for (const key of mergableKeys) {
+    if (typeof mergedNetworkSettings[key] === `undefined`) {
+      mergedNetworkSettings[key] = opts.configuration.get(key) as any;
+    }
+  }
+
+  if (process.env.NO_PROXY) {
+    const skippedHosts = process.env.NO_PROXY.split(`,`);
```
Comment (arcanis) [inline packages/yarnpkg-core/sources/httpUtils.ts:85]: `httpUtils` shouldn't rely on `process.env` - `NO_PROXY` should be "transformed" into proper settings within the configuration object.

**Review by arcanis — COMMENTED** [review]
I'd personally vote against adding support for `HTTP_PROXY` and friends, since it's duplicate with a feature we already have, and has ambiguous meaning, especially when mixing Yarn and non-Yarn setting definitions.

**Pull request merged**